### ISSUE TITLE
Missing indentation level after multiline tag

### ISSUE THREAD

**Issue by pawamoy** [rtts/djhtml#115]
Input:

```jinja
...
    {% filter heading(
        heading_level,
        role="data" if attribute.parent.kind.value == "module" else "attr",
        id=html_id,
        class="doc doc-heading",
        toc_label=('<code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;'|safe if config.show_symbol_type_toc else '') + attribute.name,
    ) %}

      {% block heading scoped %}
        ...
      {% endblock heading %}

      {% block labels scoped %}
        ...
      {% endblock labels %}

    {% endfilter %}
...
```

Output:

```jinja
...
    {% filter heading(
    heading_level,
    role="data" if attribute.parent.kind.value == "module" else "attr",
    id=html_id,
    class="doc doc-heading",
    toc_label=('<code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;'|safe if config.show_symbol_type_toc else '') + attribute.name,
    ) %}

    {% block heading scoped %}
      ...
    {% endblock heading %}

    {% block labels scoped %}
      ...
    {% endblock labels %}

  {% endfilter %}
...
```

Expected (regardless of how the opening filter tag is indented):

```jinja
...
    {% filter heading(
    heading_level,
    role="data" if attribute.parent.kind.value == "module" else "attr",
    id=html_id,
    class="doc doc-heading",
    toc_label=('<code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;'|safe if config.show_symbol_type_toc else '') + attribute.name,
    ) %}

      {% block heading scoped %}
        ...
      {% endblock heading %}
  
      {% block labels scoped %}
        ...
      {% endblock labels %}
  
    {% endfilter %}
...
```

Even better (indent continuation lines twice, and ending `) %}` only once):

```jinja
...
    {% filter heading(
        heading_level,
        role="data" if attribute.parent.kind.value == "module" else "attr",
        id=html_id,
        class="doc doc-heading",
        toc_label=('<code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;'|safe if config.show_symbol_type_toc else '') + attribute.name,
      ) %}

      {% block heading scoped %}
        ...
      {% endblock heading %}
  
      {% block labels scoped %}
        ...
      {% endblock labels %}
  
    {% endfilter %}
...
```

Reasoning: the nesting seems visually more correct, while keeping the continuation lines (`heading_level`, `role=...`, etc.) indented compared to opening and closing markers.

**Comment by pawamoy** [comment]
I'm able to at least fix detection of opening tags with this change: `OPENING_TAG = r"{%[-+]? *[#/]?(\w+).*?(?:[-+]?%})?"` (in `DjTXT`).

**Comment by pawamoy** [comment]
Here's a partial patch:

```diff
diff --git a/djhtml/modes.py b/djhtml/modes.py
index d017da3..69a5a34 100644
--- a/djhtml/modes.py
+++ b/djhtml/modes.py
@@ -56,6 +56,8 @@ class BaseMode:
         mode = self
         src = self.source
 
+        closing_marker = True
+
         while True:
             if src.find("\n") > mode.MAX_LINE_LENGTH:
                 raise MaxLineLengthExceeded
@@ -77,6 +79,14 @@ class BaseMode:
                 # Create a token from the head.
                 token, mode = mode.create_token(head, raw_token + tail, line)
                 line.append(token)
+                if isinstance(token, Token.Open):
+                    if not token.text.endswith("%}"):
+                        closing_marker = False
+                elif isinstance(token, Token.Text) and not closing_marker:
+                    if token.text.endswith("%}"):
+                        closing_marker = True
+                    else:
+                        token.relative += 1
 
             if raw_token == "\n":
                 self.lines.append(line)
@@ -209,7 +219,7 @@ class DjTXT(BaseMode):
         "video": (" as ", None),
         "placeholder": (" or ", None),
     }
-    OPENING_TAG = r"{%[-+]? *[#/]?(\w+).*?[-+]?%}"
+    OPENING_TAG = r"{%[-+]? *[#/]?(\w+).*?(?:[-+]?%})?"
 
     def create_token(self, raw_token, src, line):
         mode = self
```

Partial because it indents like this:

```jinja
{% filter hello(
        param1,
        param2,
    ) %}

    {% block world %}
        ...
    {% endblock world %}

{% endfilter %}

{% with hello(
        param1,
    param2) %}

    {{ world }}

{% endwith %}
```

First case good (`filter`), second case not so good (`with`). Not sure how to fix it.

**Comment by pawamoy** [comment]
Reproducible with `if` and `set` too:

```jinja
{%- if predicate1
    and predicate2
    and predicate3
  %}
  hello
{%- endif -%}
```

```jinja
{%- set ns = namespace(
    has_pos_only=False,
    render_pos_only_separator=True,
    render_kw_only_separator=True,
    annotation="",
    equal="=",
  ) -%}
```

(`set` is not handled by my patch above, because it's probably not detected as an opening tag)

Changed the title accordingly.

**Comment by pawamoy** [comment]
Running `djhtml` twice with my patch formats the input a first time as I expect it, then formats it again as if the patch wasn't there: I'm not sure if this is an issue with my patch or with `djhtml` more generally (non-predictable results?).

**Comment by anentropic** [comment]
Some more examples:
```jinja
        {% do item_kwargs.update({
        "checkbox_kwargs": {
        "name": _checkboxes_name,
        "value": item.value,
        "id": _checkboxes_name + "--id-" + loop.index|str,
        }
        }) %}
```
```jinja
    {{ mdc_textfield(
    name + "--filter",
    label="Filter",
    attrs={
    "@input.debounce": "(e)=>$data.filterValue=e.target.value",
    "x-init": "filteredCheckboxListFilterInit"
    }
    ) }}
```

I would love if there was something like a 'python' mode inside the multiline tags, akin to current js/css modes, so that function/macro calls could be formatted nicely, ideally with dict/list literals indented too

**Comment by JaapJoris** [comment]
With a heavy heart, I'm closing this issue as `wontfix`. If you believe that was wrong, feel free to reopen it again.

For the general case when DjHTML doesn't indent according to your preferences: you can always use the `{# fmt:off #}` and `{# fmt:on #}` tags to work around your particular issue. In this case, the following should provide some relief:

```html
{# fmt:off #}
{% filter heading(
    heading_level,
    role="data" if attribute.parent.kind.value == "module" else "attr",
    id=html_id,
    class="doc doc-heading",
    toc_label=('<code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;'|safe if config.show_symbol_type_toc else '') + attribute.name,
) %}
{# fmt:on #}
```
